icicle-compiler/src/Icicle/Runtime/Serial/Psv/Data.hs

```diff
@@ -0,0 +1,255 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+module Icicle.Runtime.Serial.Psv.Data (
+    SerialPsvDataError(..)
+  , renderSerialPsvDataError
+
+  , encodeSnapshotOutput
+  ) where
+
+import qualified Anemone.Pretty as Anemone
+
+import qualified Data.Aeson as Aeson
+import           Data.ByteString.Builder (Builder)
+import qualified Data.ByteString.Builder as Builder
+import qualified Data.ByteString.Char8 as Char8
+import qualified Data.List as List
+import qualified Data.Map.Strict as Map
+import           Data.String (String)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import qualified Data.Vector.Storable as Storable
+
+import           Icicle.Data.Name
+import           Icicle.Internal.Aeson
+import           Icicle.Runtime.Data
+import qualified Icicle.Runtime.Data.Logical as Logical
+import qualified Icicle.Runtime.Data.Schema as Schema
+import qualified Icicle.Runtime.Data.Striped as Striped
+
+import           P
+
+import           Text.Show.Pretty (ppShow)
+import           Text.Printf (printf)
+
+import qualified X.Data.Vector as Boxed
+
+import qualified Zebra.X.Vector.Segment as Segment
+
+
+data SerialPsvDataError =
+    SerialPsvDataUnsupportedSchema !Schema
+  | SerialPsvDataUnsupportedValue !Value
+  | SerialPsvDataSegmentError !Schema !Segment.SegmentError
+  | SerialPsvDataStripedError !Striped.StripedError
+  | SerialPsvDataOutputError !OutputId !SerialPsvDataError
+  | SerialPsvDataMissingValue !Error64
+    deriving (Eq, Show)
+
+renderSerialPsvDataError :: SerialPsvDataError -> Text
+renderSerialPsvDataError = \case
+  SerialPsvDataUnsupportedSchema x ->
+    "Unsupported PSV output: " <> Text.pack (ppShow x)
+  SerialPsvDataUnsupportedValue x ->
+    "Unsupported PSV output: " <> Text.pack (ppShow x)
+  SerialPsvDataSegmentError s x ->
+    "Error encoding PSV, invalid segment descriptor for " <> Text.pack (ppShow s) <> ": " <> Segment.renderSegmentError x
+  SerialPsvDataStripedError x ->
+    "Error encoding PSV: " <> Striped.renderStripedError x
+  SerialPsvDataOutputError oid x ->
+    "Error encoding <" <> renderOutputId oid <> "> output: " <> renderSerialPsvDataError x
+  SerialPsvDataMissingValue x ->
+    "Internal error, found missing value: " <> Text.pack (ppShow x)
+
+encodeBool :: Bool64 -> Builder
+encodeBool = \case
+  False64 ->
+    "false"
+  _ ->
+    "true"
+{-# INLINE encodeBool #-}
+
+encodeInt :: Int64 -> Builder
+encodeInt =
+  Builder.int64Dec
+{-# INLINE encodeInt #-}
+
+encodeDouble :: Double -> Builder
+encodeDouble =
+  Builder.byteString . Anemone.renderDouble
+{-# INLINE encodeDouble #-}
+
+encodeDateString :: Time64 -> String
+encodeDateString x =
+  let
+    UnpackedTime64 y m d _ =
+      unpackTime x
+  in
+    printf "%04d-%02d-%02d" y m d
```

Comment:
yeah, it's good to be explicit about this. We don't actually output time in PSV.